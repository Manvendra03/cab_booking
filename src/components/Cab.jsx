import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { FaMediumM, FaTwitter, FaInstagram } from 'react-icons/fa';

const Cab = ({ cab }) => (
  <li
    key={cab.model}
    className="grid grid-rows-2 items-start gap-4 transition-all duration-300 ease-linear"
  >
    <Link to={`/cabs/${cab.id}`} className="flex flex-col gap-4 items-center">
      <img
        src={cab.image_url}
        alt={cab.model}
        className="rounded-full aspect-square w-40 "
      />
    </Link>
    <section className="flex flex-col gap-4 items-center justify-evenly">
      <h2 className="font-extrabold text-[#645858] text-xl tracking-wider">
        {`${cab.model.substring(0, 15)}`}
      </h2>
      <hr className="border-dotted w-full font-bold text-2xl border-[3px]" />
      <p className="text-[#807e80]">
        {`${cab.description.substring(0, 100)}...`}
      </p>
      <div className="flex gap-4">
        <span className="rounded-full border p-2">
          <FaTwitter className="text-xl  text-[#a7a4a7]  cursor-pointer  " />
        </span>
        <span className="rounded-full border p-2">
          <FaInstagram className="text-xl text-[#a7a4a7] cursor-pointer  " />
        </span>
        <span className="rounded-full border p-2">
          <FaMediumM className="text-xl  text-[#a7a4a7]  cursor-pointer  " />
        </span>
      </div>
    </section>
  </li>
);

export default Cab;

Cab.propTypes = {
  cab: PropTypes.shape({
    model: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    image_url: PropTypes.string.isRequired,
    id: PropTypes.number.isRequired,
  }).isRequired,
};
