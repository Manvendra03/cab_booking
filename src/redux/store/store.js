import { configureStore } from '@reduxjs/toolkit';
import userReducer from '../slices/userSlice';
import fetchCabReducer from '../slices/cabSlice';
import reservationReducer from '../slices/reservationSlice';
import menuReducer from '../slices/menuSlice';

const store = configureStore({
  reducer: {
    fetchCab: fetchCabReducer,
    user: userReducer,
    reservation: reservationReducer,
    menu: menuReducer,
  },
});

export default store;
