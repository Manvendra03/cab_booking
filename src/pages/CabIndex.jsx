import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { BiLeftArrow, BiRightArrow } from 'react-icons/bi';
import { motion } from 'framer-motion';
import { Circles } from 'react-loader-spinner';
import { nextCab, prevCab, fetchCabs } from '../redux/slices/cabSlice';
import Cab from '../components/Cab';

const CabIndex = () => {
  const [cabsLists, setCabsLists] = useState([]);
  const { loading } = useSelector((state) => state.fetchCab);
  const dispatch = useDispatch();
  const cabs = useSelector((state) => state.fetchCab.displayedCabs);
  const currentPage = useSelector((state) => state.fetchCab.currentPage);
  const numOfPages = useSelector((state) => state.fetchCab.numOfPages);

  const totalCabs = useSelector((state) => state.fetchCab.cabs);

  useEffect(() => {
    dispatch(fetchCabs());
  }, [dispatch]);
  useEffect(() => {
    setCabsLists(cabs);
  }, [cabs]);

  if (loading) {
    return (
      <div className="h-screen flex justify-center items-center">
        <Circles
          height="80"
          width="80"
          color="rgba(101, 163, 13, 1)"
          ariaLabel="circles-loading"
          wrapperStyle={{}}
          wrapperClass=""
          visible={loading}
        />
      </div>
    );
  }

  return (
    <main className="py-20 transition-all h-screen overflow-auto duration-300 ease-linear">
      <header className="flex flex-col gap-2 md:gap-4 text-lime-800 items-center">
        <h1 className="font-extrabold text-2xl lg:text-3xl xl:text-5xl capitalize">
          Secure your ride
        </h1>
        <p className="text-sm md:text-base font-semibold italic ">
          Please select a cab for rentals
        </p>
        <hr className="border-dotted w-[300px] mt-4 font-bold text-2xl border-[3px]" />
      </header>
      {cabsLists.length === 0 && (
        <section className="flex justify-center pt-20 text-lime-800 items-center">
          <h1 className="text-2xl text-center font-bold tracking-widest">
            No cabs available...
          </h1>
        </section>
      )}
      <section className="mb-16 lg:mb-0">
        <ul className="grid pb-12 grid-cols-1 lg:grid-cols-2 xl:grid-cols-3 relative gap-8 gap-y-16  items-center transition-all duration-300 ease-linear justify-center p-12 px-4 md:px-12 lg:px-20 xl:px-32">
          {totalCabs.length > 3 && (
            <>
              <button
                className={`absolute top-full my-8 lg:top-1/2 left-0 transform -translate-y-1/2 ${
                  currentPage > 1 ? 'bg-lime-400' : 'bg-slate-200'
                } p-6 lg:p-8 pl-10 lg:pl-16 rounded-r-full`}
                type="button"
                onClick={() => dispatch(prevCab())}
                disabled={currentPage === 1}
              >
                <BiLeftArrow className="text-white text-xl" />
              </button>
              <button
                className={`absolute top-full my-8 lg:top-1/2 right-0 transform -translate-y-1/2 ${
                  currentPage === numOfPages - 2
                    ? 'bg-slate-200'
                    : 'bg-lime-400'
                } p-6 lg:p-8 pr-10 lg:pr-16 rounded-l-full`}
                type="button"
                onClick={() => dispatch(nextCab())}
                disabled={currentPage === numOfPages - 2}
              >
                <BiRightArrow className="text-white text-xl" />
              </button>
            </>
          )}
          {cabsLists.map((cab) => (
            <motion.div key={cab.id} className="shadow-xl p-4 lg:shadow-none">
              <Cab cab={cab} />
            </motion.div>
          ))}
        </ul>
      </section>
    </main>
  );
};

export default CabIndex;
