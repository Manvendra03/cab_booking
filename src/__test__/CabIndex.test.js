import { render, screen, waitFor } from '@testing-library/react';
import TestRenderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import CabIndex from '../pages/CabIndex';
import cabData from '../library/cabs';
import store from '../redux/store/store';
import Cab from '../components/Cab';

/**
 * @jest-environment jsdom
 */

const mock = new MockAdapter(axios);
const mockStore = configureMockStore([thunk]);

describe('CabIndex', () => {
  it('renders', () => {
    mock.onGet('http://127.0.0.1:3000/api/v1/cabs').reply(200, cabData);
    const tree = TestRenderer.create(
      <Provider store={mockStore(store.getState())}>
        <BrowserRouter>
          <CabIndex />
        </BrowserRouter>
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders CabIndex and checks for main header', async () => {
    mock.onGet('http://127.0.0.1:3000/api/v1/cabs').reply(200, cabData);
    render(
      <Provider store={mockStore(store.getState())}>
        <BrowserRouter>
          <CabIndex />
        </BrowserRouter>
      </Provider>,
    );

    await waitFor(() => {
      const header = screen.getByText(/Secure your ride/i);
      expect(header).toBeInTheDocument();
    });
  });
  it('renders CabIndex and checks for the subheader', async () => {
    mock.onGet('http://127.0.0.1:3000/api/v1/cabs').reply(200, cabData);

    render(
      <Provider store={mockStore(store.getState())}>
        <BrowserRouter>
          <CabIndex />
        </BrowserRouter>
      </Provider>,
    );

    await waitFor(() => {
      const subHead = screen.getByText(/Please select a cab for rentals/i);
      expect(subHead).toBeInTheDocument();
    });
  });

  test('renders CabIndex and checks for existence of data', async () => {
    mock.onGet('http://127.0.0.1:3000/api/v1/cabs').reply(200, cabData);

    render(
      <Provider store={mockStore(store.getState())}>
        <BrowserRouter>
          <Cab cab={cabData[0]} />
        </BrowserRouter>
      </Provider>,
    );

    // Wait for the component to update with fetched data
    await waitFor(() => {
      const dataElement = screen.getByText(cabData[0].model);
      expect(dataElement).toBeInTheDocument();
    });
  });
});
