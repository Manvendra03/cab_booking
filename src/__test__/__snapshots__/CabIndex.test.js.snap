// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CabIndex renders 1`] = `
<main
  className="py-20 transition-all h-screen overflow-auto duration-300 ease-linear"
>
  <header
    className="flex flex-col gap-2 md:gap-4 text-lime-800 items-center"
  >
    <h1
      className="font-extrabold text-2xl lg:text-3xl xl:text-5xl capitalize"
    >
      Secure your ride
    </h1>
    <p
      className="text-sm md:text-base font-semibold italic "
    >
      Please select a cab for rentals
    </p>
    <hr
      className="border-dotted w-[300px] mt-4 font-bold text-2xl border-[3px]"
    />
  </header>
  <section
    className="flex justify-center pt-20 text-lime-800 items-center"
  >
    <h1
      className="text-2xl text-center font-bold tracking-widest"
    >
      No cabs available...
    </h1>
  </section>
  <section
    className="mb-16 lg:mb-0"
  >
    <ul
      className="grid pb-12 grid-cols-1 lg:grid-cols-2 xl:grid-cols-3 relative gap-8 gap-y-16  items-center transition-all duration-300 ease-linear justify-center p-12 px-4 md:px-12 lg:px-20 xl:px-32"
    />
  </section>
</main>
`;
